File: android/src/main/java/com/backendless/backendless_sdk/call_handlers/GeoCallHandler.java
package com.backendless.backendless_sdk.call_handlers;

import android.util.SparseArray;

import com.backendless.Backendless;
import com.backendless.geo.BackendlessGeoQuery;
import com.backendless.geo.GeoCategory;
import com.backendless.geo.GeoCluster;
import com.backendless.geo.GeoPoint;
import com.backendless.geo.SearchMatchesResult;
import com.backendless.geo.geofence.IGeofenceCallback;
import com.backendless.backendless_sdk.utils.FlutterCallback;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class GeoCallHandler implements MethodChannel.MethodCallHandler {
    private MethodChannel methodChannel;
    private final SparseArray<GeoFenceEventCallback> geofenceCallbacks = new SparseArray<>();

    public GeoCallHandler(MethodChannel methodChannel) {
        this.methodChannel = methodChannel;
    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        switch (call.method) {
            case "Backendless.Geo.addCategory":
                addCategory(call, result);
                break;
            case "Backendless.Geo.deleteCategory":
                deleteCategory(call, result);
                break;
            case "Backendless.Geo.getCategories":
                getCategories(result);
                break;
            case "Backendless.Geo.getGeopointCount":
                getGeopointCount(call, result);
                break;
            case "Backendless.Geo.getPoints":
                getPoints(call, result);
                break;
            case "Backendless.Geo.loadMetadata":
                loadMetadata(call, result);
                break;
            case "Backendless.Geo.relativeFind":
                relativeFind(call, result);
                break;
            case "Backendless.Geo.removePoint":
                removePoint(call);
                break;
            case "Backendless.Geo.runOnEnterAction":
                runOnEnterAction(call);
                break;
            case "Backendless.Geo.runOnExitAction":
                runOnExitAction(call);
                break;
            case "Backendless.Geo.runOnStayAction":
                runOnStayAction(call);
                break;
            case "Backendless.Geo.savePoint":
                savePoint(call, result);
                break;
            case "Backendless.Geo.setLocationTrackerParameters":
                setLocationTrackerParameters(call);
                break;
            case "Backendless.Geo.startClientGeofenceMonitoring":
                startClientGeofenceMonitoring(call, result);
                break;
            case "Backendless.Geo.startServerGeofenceMonitoring":
                startServerGeofenceMonitoring(call, result);
                break;
            case "Backendless.Geo.stopGeofenceMonitoring":
                stopGeofenceMonitoring(call, result);
                break;
            default:
                result.notImplemented();
        }
    }

    private void addCategory(MethodCall call, MethodChannel.Result result) {
        String categoryName = call.argument("categoryName");
        Backendless.Geo.addCategory(categoryName, new FlutterCallback<GeoCategory>(result));
    }

    private void deleteCategory(MethodCall call, MethodChannel.Result result) {
        String categoryName = call.argument("categoryName");
        Backendless.Geo.deleteCategory(categoryName, new FlutterCallback<Boolean>(result));
    }

    private void getCategories(MethodChannel.Result result) {
        Backendless.Geo.getCategories(new FlutterCallback<List<GeoCategory>>(result));
    }

    private void getGeopointCount(MethodCall call, MethodChannel.Result result) {
        BackendlessGeoQuery query = call.argument("query");
        String geoFenceName = call.argument("geoFenceName");

        FlutterCallback<Integer> callback = new FlutterCallback<>(result);

        if (geoFenceName != null) {
            Backendless.Geo.getGeopointCount(geoFenceName, query, callback);
        } else {
            Backendless.Geo.getGeopointCount(query, callback);
        }
    }

    private void getPoints(MethodCall call, MethodChannel.Result result) {
        BackendlessGeoQuery query = call.argument("query");
        String geofenceName = call.argument("geofenceName");
        GeoCluster geoCluster = call.argument("geoCluster");

        FlutterCallback<List<GeoPoint>> callback = new FlutterCallback<>(result);

        if (geoCluster != null) {
            Backendless.Geo.getPoints(geoCluster, callback);
        } else if (geofenceName != null) {
            if (query != null) {
                Backendless.Geo.getPoints(geofenceName, query, callback);
            } else {
                Backendless.Geo.getPoints(geofenceName, callback);
            }
        } else if (query != null) {
            Backendless.Geo.getPoints(query, callback);
        }
    }

    private void loadMetadata(MethodCall call, MethodChannel.Result result) {
        GeoPoint geoPoint = call.argument("geoPoint");
        Backendless.Geo.loadMetadata(geoPoint, new FlutterCallback<GeoPoint>(result));
    }

    private void relativeFind(MethodCall call, MethodChannel.Result result) {
        BackendlessGeoQuery geoQuery = call.argument("geoQuery");
        Backendless.Geo.relativeFind(geoQuery, new FlutterCallback<List<SearchMatchesResult>>(result));
    }

    private void removePoint(MethodCall call) {
        GeoPoint geoPoint = call.argument("geoPoint");
        Backendless.Geo.removePoint(geoPoint, new FlutterCallback<Void>(null));
    }

    private void runOnEnterAction(MethodCall call) {
        String geoFenceName = call.argument("geoFenceName");
        GeoPoint geoPoint = call.argument("geoPoint");

        if (geoPoint != null) {
            Backendless.Geo.runOnEnterAction(geoFenceName, geoPoint, new FlutterCallback<Void>(null));
        } else {
            Backendless.Geo.runOnEnterAction(geoFenceName, new FlutterCallback<Integer>(null));
        }
    }

    private void runOnExitAction(MethodCall call) {
        String geoFenceName = call.argument("geoFenceName");
        GeoPoint geoPoint = call.argument("geoPoint");

        if (geoPoint != null) {
            Backendless.Geo.runOnExitAction(geoFenceName, geoPoint, new FlutterCallback<Void>(null));
        } else {
            Backendless.Geo.runOnExitAction(geoFenceName, new FlutterCallback<Integer>(null));
        }
    }

    private void runOnStayAction(MethodCall call) {
        String geoFenceName = call.argument("geoFenceName");
        GeoPoint geoPoint = call.argument("geoPoint");

        if (geoPoint != null) {
            Backendless.Geo.runOnStayAction(geoFenceName, geoPoint, new FlutterCallback<Void>(null));
        } else {
            Backendless.Geo.runOnStayAction(geoFenceName, new FlutterCallback<Integer>(null));
        }
    }

    private void savePoint(MethodCall call, MethodChannel.Result result) {
        FlutterCallback<GeoPoint> callback = new FlutterCallback<>(result);

        GeoPoint geoPoint = call.argument("geoPoint");
        if (geoPoint != null) {
            Backendless.Geo.savePoint(geoPoint, callback);
        } else {
            Double latitude = call.argument("latitude");
            Double longitude = call.argument("longitude");
            List<String> categories = call.argument("categories");
            Map<String, Object> metadata = call.argument("metadata");

            if (categories != null) {
                Backendless.Geo.savePoint(latitude, longitude, categories, metadata, callback);
            } else {
                Backendless.Geo.savePoint(latitude, longitude, metadata, callback);
            }
        }
    }

    private void setLocationTrackerParameters(MethodCall call) {
        Integer minTime = call.argument("minTime");
        Integer minDistance = call.argument("minDistance");
        Integer acceptedDistanceAfterReboot = call.argument("acceptedDistanceAfterReboot");

        Backendless.Geo.setLocationTrackerParameters(minTime, minDistance, acceptedDistanceAfterReboot);
    }

    private void startClientGeofenceMonitoring(MethodCall call, MethodChannel.Result result) {
        String geofenceName = call.argument("geofenceName");
        int handle = call.argument("handle");
        GeoFenceEventCallback callback = new GeoFenceEventCallback(handle);
        geofenceCallbacks.put(handle, callback);
        if (geofenceName != null) {
            Backendless.Geo.startGeofenceMonitoring(geofenceName, callback, new FlutterCallback<Void>(result));
        } else {
            Backendless.Geo.startGeofenceMonitoring(callback, new FlutterCallback<Void>(result));
        }
    }

    private void startServerGeofenceMonitoring(MethodCall call, MethodChannel.Result result) {
        GeoPoint geoPoint = call.argument("geoPoint");
        String geofenceName = call.argument("geofenceName");

        if (geofenceName != null) {
            Backendless.Geo.startGeofenceMonitoring(geofenceName, geoPoint, new FlutterCallback<Void>(result));
        } else {
            Backendless.Geo.startGeofenceMonitoring(geoPoint, new FlutterCallback<Void>(result));
        }
    }

    private void stopGeofenceMonitoring(MethodCall call, MethodChannel.Result result) {
        String geofenceName = call.argument("geofenceName");

        if (geofenceName != null) {
            Backendless.Geo.stopGeofenceMonitoring(geofenceName);
        } else {
            Backendless.Geo.stopGeofenceMonitoring();
        }
        result.success(null);
    }

    class GeoFenceEventCallback implements IGeofenceCallback {
        private int handle;

        public GeoFenceEventCallback(int handle) {
            this.handle = handle;
        }

        @Override
        public void geoPointEntered(String geofenceName, String geofenceId, double latitude, double longitude) {
            invokeMethod(geofenceName, geofenceId, latitude, longitude, "geoPointEntered");
        }

        @Override
        public void geoPointStayed(String geofenceName, String geofenceId, double latitude, double longitude) {
            invokeMethod(geofenceName, geofenceId, latitude, longitude, "geoPointStayed");
        }

        @Override
        public void geoPointExited(String geofenceName, String geofenceId, double latitude, double longitude) {
            invokeMethod(geofenceName, geofenceId, latitude, longitude, "geoPointExited");
        }

        private void invokeMethod(String geofenceName, String geofenceId, double latitude, double longitude, String method) {
            Map<String, Object> arguments = new HashMap<>();
            arguments.put("handle", handle);
            arguments.put("geofenceName", geofenceName);
            arguments.put("geofenceId", geofenceId);
            arguments.put("latitude", latitude);
            arguments.put("longitude", longitude);
            arguments.put("method", method);
            methodChannel.invokeMethod("Backendless.Geo.GeofenceMonitoring", arguments);
        }
    }
}

File: ios/Classes/CallHandlers/GeoCallHandler.swift
//
//  GeoCallHandler.swift
//  Flutter-SDK
//
//  Created by Andrii Bodnar on 5/2/19.
//  Copyright © 2019 Andrii Bodnar. All rights reserved.
//

import Foundation
import Flutter
import Backendless

class GeoCallHandler: FlutterCallHandlerProtocol {
    
    // MARK: -
    // MARK: - Constants
    private enum Methods {
        static let addCategory = "Backendless.Geo.addCategory"
        static let deleteCategory = "Backendless.Geo.deleteCategory"
        static let getCategories = "Backendless.Geo.getCategories"
        static let getGeopointCount = "Backendless.Geo.getGeopointCount"
        static let getPoints = "Backendless.Geo.getPoints"
        static let loadMetadata = "Backendless.Geo.loadMetadata"
        static let relativeFind = "Backendless.Geo.relativeFind"
        static let removePoint = "Backendless.Geo.removePoint"
        static let runOnEnterAction = "Backendless.Geo.runOnEnterAction"
        static let runOnExitAction = "Backendless.Geo.runOnExitAction"
        static let runOnStayAction = "Backendless.Geo.runOnStayAction"
        static let savePoint = "Backendless.Geo.savePoint"
        static let setLocationTrackerParameters = "Backendless.Geo.setLocationTrackerParameters"
        static let startClientGeofenceMonitoring = "Backendless.Geo.startClientGeofenceMonitoring"
        static let startServerGeofenceMonitoring = "Backendless.Geo.startServerGeofenceMonitoring"
        static let stopGeofenceMonitoring = "Backendless.Geo.stopGeofenceMonitoring"
    }
    
    private enum Args {
        static let categoryName = "categoryName"
        static let query = "query"
        static let geofenceName = "geoFenceName"
        static let geoPoint = "geoPoint"
        static let latitude = "latitude"
        static let longitude = "longitude"
        static let categories = "categories"
        static let metadata = "metadata"
        static let geoCluster = "geoCluster"
    }
    
    // MARK: -
    // MARK: - Geo Reference
    private let geo = SwiftBackendlessSdkPlugin.backendless.geo
    
    // MARK: -
    // MARK: - Route Flutter Call
    func routeFlutterCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let arguments: [String: Any] = call.arguments.flatMap(cast) ?? [:]
        
        switch call.method {
        case Methods.addCategory:
            addCategory(arguments, result)
        case Methods.deleteCategory:
            deleteCategory(arguments, result)
        case Methods.getCategories:
            getCategories(arguments, result)
        case Methods.getGeopointCount:
            getGeopointCount(arguments, result)
        case Methods.getPoints:
            getPoints(arguments, result)
        case Methods.loadMetadata:
            loadMetadata(arguments, result)
        case Methods.relativeFind:
            relativeFind(arguments, result)
        case Methods.removePoint:
            removePoint(arguments, result)
        case Methods.runOnEnterAction:
            runOnEnterAction(arguments, result)
        case Methods.runOnExitAction:
            runOnExitAction(arguments, result)
        case Methods.runOnStayAction:
            runOnStayAction(arguments, result)
        case Methods.savePoint:
            savePoint(arguments, result)
        case Methods.setLocationTrackerParameters:
            setLocationTrackerParameters(arguments, result)
        case Methods.startClientGeofenceMonitoring:
            startClientGeofenceMonitoring(arguments, result)
        case Methods.startServerGeofenceMonitoring:
            startServerGeofenceMonitoring(arguments, result)
        case Methods.stopGeofenceMonitoring:
            stopGeofenceMonitoring(arguments, result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: -
    // MARK: - AddCategory
    private func addCategory(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let categoryName: String = arguments[Args.categoryName].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        geo.addCategory(categoryName: categoryName,
            responseHandler: { (category: GeoCategory) in
                result(category)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - DeleteCategory
    private func deleteCategory(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let categoryName: String = arguments[Args.categoryName].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        geo.deleteGeoCategory(categoryName: categoryName,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - GetCategories
    private func getCategories(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        geo.getCategories(responseHandler: {
            result($0)
        }, errorHandler: {
            result(FlutterError($0))
        })
    }
    
    // MARK: -
    // MARK: - GetGeopointCount
    private func getGeopointCount(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        
        // TODO: -
        // TODO: - Need to test: problem with getting BackendlessGeoQuery object from Flutter
        
        // TODO: -
        // TODO: - How to use GeofenceName?
        
        let geoQuery: BackendlessGeoQuery? = arguments[Args.query].flatMap(cast)
        let geofenceName: String? = arguments[Args.geofenceName].flatMap(cast)
        
        if let geoQuery = geoQuery {
            geo.getPointsCount(geoQuery: geoQuery,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.getPointsCount(responseHandler: {
                result($0)
            }, errorHandler: {
                result(FlutterError($0))
            })
        }
    }
    
    // MARK: -
    // MARK: - GetPoints +-
    private func getPoints(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        
        // TODO: -
        // TODO: - How to use GeofenceName & GeoCluster?
        
        let geoQuery: BackendlessGeoQuery? = arguments[Args.query].flatMap(cast)
        let geofenceName: String? = arguments[Args.geofenceName].flatMap(cast)
        let geoCluster: GeoCluster? = arguments[Args.geoCluster].flatMap(cast)
        
        if let query = geoQuery {
            geo.getPoints(geoQuery: query,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.getPoints(responseHandler: {
                result($0)
            }, errorHandler: {
                result(FlutterError($0))
            })
        }
    }
    
    // MARK: -
    // MARK: - Load Metadata
    private func loadMetadata(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geoPoint: GeoPoint = arguments[Args.geoPoint].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        geo.loadMetadata(geoPoint: geoPoint,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - RelativeFind
    private func relativeFind(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        
        // TODO: -
        // TODO: - Need to test. Problems with getting BackendlessGeoQuery from Flutter
        
        guard let geoQuery: BackendlessGeoQuery = arguments[Args.query].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        geo.relativeFind(geoQuery: geoQuery,
            responseHandler: {
                result($0)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - RemovePoint
    private func removePoint(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geoPoint: GeoPoint = arguments[Args.geoPoint].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        geo.removeGeoPoint(geoPoint: geoPoint,
            responseHandler: {
                result(nil)
            },
            errorHandler: {
                result(FlutterError($0))
            })
    }
    
    // MARK: -
    // MARK: - RunOnEnterAction
    private func runOnEnterAction(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geofenceName: String = arguments[Args.geofenceName].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        let geoPoint: GeoPoint? = arguments[Args.geoPoint].flatMap(cast)
        
        if let geoPoint = geoPoint {
            geo.runOnEnterAction(geoFenceName: geofenceName, geoPoint: geoPoint,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.runOnEnterAction(geoFenceName: geofenceName,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        }
    }
    
    // MARK: -
    // MARK: - RunOnExitAction
    private func runOnExitAction(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geofenceName: String = arguments[Args.geofenceName].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        let geoPoint: GeoPoint? = arguments[Args.geoPoint].flatMap(cast)
        
        if let geoPoint = geoPoint {
            geo.runOnExitAction(geoFenceName: geofenceName, geoPoint: geoPoint,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.runOnExitAction(geoFenceName: geofenceName,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        }
    }
    
    // MARK: -
    // MARK: - RunOnStayAction
    private func runOnStayAction(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geofenceName: String = arguments[Args.geofenceName].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        let geoPoint: GeoPoint? = arguments[Args.geoPoint].flatMap(cast)
        
        if let geoPoint = geoPoint {
            geo.runOnStayAction(geoFenceName: geofenceName, geoPoint: geoPoint,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.runOnStayAction(geoFenceName: geofenceName,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        }
    }
    
    // MARK: -
    // MARK: - SavePoint
    private func savePoint(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        let geoPoint: GeoPoint? = arguments[Args.geoPoint].flatMap(cast)
        
        if let geoPoint = geoPoint {
            geo.saveGeoPoint(geoPoint: geoPoint,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            let newGeoPoint = GeoPoint()
            
            arguments[Args.latitude]
                .flatMap(cast)
                .flatMap { newGeoPoint.latitude = $0 }
            
            arguments[Args.longitude]
                .flatMap(cast)
                .flatMap { newGeoPoint.longitude = $0 }
            
            arguments[Args.categories]
                .flatMap(cast)
                .flatMap { newGeoPoint.categories = $0 }
            
            arguments[Args.metadata]
                .flatMap(cast)
                .flatMap { newGeoPoint.metadata = $0 }
            
            geo.saveGeoPoint(geoPoint: newGeoPoint,
                responseHandler: {
                    result($0)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        }
    }
    
    // MARK: -
    // MARK: - SetLocationTrackerParameters
    private func setLocationTrackerParameters(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        
        // TODO: -
        // TODO: - No such method in iOS SDK
        
        result(FlutterMethodNotImplemented)
    }
    
    // MARK: -
    // MARK: - StartClientGeofenceMonitoring
    private func startClientGeofenceMonitoring(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        
        // TODO: -
        // TODO: - Using IGeofenceCallback
        
    }
    
    // MARK: -
    // MARK: - StartServerGeofenceMonitoring
    private func startServerGeofenceMonitoring(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let geoPoint: GeoPoint = arguments[Args.geoPoint].flatMap(cast) else {
            result(FlutterError.noRequiredArguments)
            
            return
        }
        
        let geofenceName: String? = arguments[Args.geofenceName].flatMap(cast)
        
        if let geofenceName = geofenceName {
            geo.startGeoFenceMonitoring(geoFenceName: geofenceName, geoPoint: geoPoint,
                responseHandler: {
                    result(nil)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        } else {
            geo.startGeoFenceMonitoring(geoPoint: geoPoint,
                responseHandler: {
                    result(nil)
                },
                errorHandler: {
                    result(FlutterError($0))
                })
        }
    }
    
    // MARK: -
    // MARK: - StopGeofenceMonitoring
    private func stopGeofenceMonitoring(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        let geofenceName: String? = arguments[Args.geofenceName].flatMap(cast)
        
        if let geofenceName = geofenceName {
            geo.stopGeoFenceMonitoring(geoFenceName: geofenceName)
        } else {
            geo.stopGeoFenceMonitoring()
        }
        
        result(nil)
    }
    
    
}
